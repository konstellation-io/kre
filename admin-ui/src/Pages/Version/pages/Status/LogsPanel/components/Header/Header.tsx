import * as ICONS from 'Constants/icons';

import IconClose from '@material-ui/icons/ExpandMore';
import IconExpand from '@material-ui/icons/Fullscreen';
import IconExpandExit from '@material-ui/icons/FullscreenExit';
import IconOpen from '@material-ui/icons/ExpandLess';
import React from 'react';
import SvgIcon from '@material-ui/core/SvgIcon';
import cx from 'classnames';
import styles from './Header.module.scss';

type Props = {
  togglePanel: () => void;
  toggleFullScreen: () => void;
  opened: boolean;
  fullScreen: boolean;
};
function Header({ togglePanel, toggleFullScreen, opened, fullScreen }: Props) {
  const IconOpenLogs = opened ? IconClose : IconOpen;
  const titleOpenLogs = opened ? 'Close logs panel' : 'Open logs panel';
  const IconFullScreen = fullScreen ? IconExpandExit : IconExpand;
  const titleFullScreen = fullScreen ? 'Contract' : 'Expand';

  return (
    <div
      className={cx(styles.container, {
        [styles.opened]: opened
      })}
      onClick={() => !opened && togglePanel()}
      title="Open logs"
    >
      <div className={styles.title}>
        <SvgIcon className="icon-small">
          <path d={ICONS.TERMINAL} />
        </SvgIcon>
        <span>LOGS CONSOLE</span>
      </div>
      <div className={styles.buttons}>
        {opened && (
          <div onClick={toggleFullScreen} title={titleFullScreen}>
            <IconFullScreen className="icon-regular" />
          </div>
        )}
        <div onClick={togglePanel} title={titleOpenLogs}>
          <IconOpenLogs className="icon-regular" />
        </div>
      </div>
    </div>
  );
}

export default Header;
