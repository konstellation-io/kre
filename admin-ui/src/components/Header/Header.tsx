import { useQuery } from '@apollo/react-hooks';
import { loader } from 'graphql.macro';
import { get } from 'lodash';
import React, { FunctionComponent, ReactElement } from 'react';
import Settings from '../../components/Settings/Settings';
import { GetMe } from '../../graphql/queries/types/GetMe';
import styles from './Header.module.scss';

const GetMeQuery = loader('../../graphql/queries/getMe.graphql');

type Props = {
  children?: ReactElement | ReactElement[] | null;
  hideSettings?: boolean;
};
const Header: FunctionComponent<Props> = ({
  children,
  hideSettings = false
}) => {
  const { data, loading } = useQuery<GetMe>(GetMeQuery);

  if (loading)
    return <div className={styles.splash} data-testid="splashscreen" />;

  const username: string = get(data, 'me.email');

  return (
    <header className={styles.container} data-testid="app-header">
      <img
        className={styles.konstellationText}
        src={'/img/brand/konstellation.svg'}
        alt="konstellation"
      />
      <div className={styles.customHeaderElements}>{children}</div>
      {!hideSettings && <Settings label={username} />}
    </header>
  );
};

export default Header;
