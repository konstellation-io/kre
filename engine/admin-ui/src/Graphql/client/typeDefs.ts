import { AccessLevel, LogLevel } from 'Graphql/types/globalTypes';

export interface NodeSelection {
  workflowName: string;
  nodeNames: string[];
  __typename: string;
}

export interface LogPanelFilters {
  dateOption?: string;
  startDate?: string;
  endDate?: string;
  nodes?: NodeSelection[];
  search?: string;
  levels?: LogLevel[] | null;
  versionsIds?: string[] | null;
  workflowsNames?: string[] | null;
  __typename: 'logTabFilters';
}

export interface LogPanel {
  runtimeId: string;
  runtimeName: string;
  versionId: string;
  versionName: string;
  uniqueId: string;
  filters: LogPanelFilters;
}

export interface UserSettingsFilters {
  email: string | null;
  accessLevel: AccessLevel | null;
  __typename: 'UserSettingsFilters';
}
