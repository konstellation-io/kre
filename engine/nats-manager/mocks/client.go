// Code generated by MockGen. DO NOT EDIT.
// Source: client.go

// Package mocks is a generated GoMock package.
package mocks
