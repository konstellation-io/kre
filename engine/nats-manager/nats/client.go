package nats

import (
	"fmt"
	"regexp"

	"github.com/nats-io/nats.go"

	"github.com/konstellation-io/kre/engine/nats-manager/internal/entity"
	"github.com/konstellation-io/kre/engine/nats-manager/internal/errors"
	"github.com/konstellation-io/kre/engine/nats-manager/internal/logging"
)

type NatsClient struct {
	js     nats.JetStreamContext
	logger logging.Logger
}

func New(logger logging.Logger, js nats.JetStreamContext) *NatsClient {
	return &NatsClient{
		logger: logger,
		js:     js,
	}
}

func InitJetStreamConnection(url string) (nats.JetStreamContext, error) {
	natsConn, err := nats.Connect(url)
	if err != nil {
		return nil, fmt.Errorf("error connecting to NATS: %w", err)
	}

	js, err := natsConn.JetStream()
	if err != nil {
		return nil, fmt.Errorf("error connecting to NATS JetStream: %w", err)
	}

	return js, nil
}

func (n *NatsClient) CreateStream(streamConfig *entity.StreamConfig) error {
	n.logger.Infof("Creating stream  %q", streamConfig.Stream)

	subjects := n.getNodesSubjects(streamConfig.Nodes)

	streamCfg := &nats.StreamConfig{
		Name:        streamConfig.Stream,
		Description: "",
		Subjects:    append(subjects, streamConfig.EntrypointSubject),
		Retention:   nats.InterestPolicy,
	}

	_, err := n.js.AddStream(streamCfg)
	return err
}

func (n *NatsClient) CreateObjectStores(objectStore string) error {
	n.logger.Infof("Creating object store %q", objectStore)

	_, err := n.js.CreateObjectStore(&nats.ObjectStoreConfig{
		Bucket:  objectStore,
		Storage: nats.FileStorage,
	})
	if err != nil {
		return fmt.Errorf("error creating the object store: %w", err)
	}

	return nil
}

// GetObjectStoreNames returns the list of object stores.
// The optional param `optFilter` accepts 0 or 1 value.
func (n *NatsClient) GetObjectStoreNames(optFilter ...*regexp.Regexp) ([]string, error) {
	if len(optFilter) > 1 {
		return nil, errors.ErrNoOptFilter
	}

	var regexpFilter *regexp.Regexp
	if len(optFilter) == 1 {
		regexpFilter = optFilter[0]
	}

	objectStoresCh := n.js.ObjectStores()
	objectStores := make([]string, 0)

	for objectStore := range objectStoresCh {
		objStoreName := objectStore.Bucket()

		nameMatchFilter := regexpFilter == nil || regexpFilter.MatchString(objStoreName)
		if nameMatchFilter {
			objectStores = append(objectStores, objStoreName)
		}
	}

	return objectStores, nil
}

func (n *NatsClient) CreateObjectStore(objectStore string) error {
	n.logger.Infof("Creating object store %q", objectStore)

	_, err := n.js.CreateObjectStore(&nats.ObjectStoreConfig{
		Bucket:  objectStore,
		Storage: nats.FileStorage,
	})
	if err != nil {
		return fmt.Errorf("error creating the object store: %w", err)
	}

	return nil
}

func (n *NatsClient) CreateKeyValueStore(keyValueStore string) error {
	n.logger.Infof("Creating key-value store %q", keyValueStore)

	_, err := n.js.CreateKeyValue(&nats.KeyValueConfig{
		Bucket: keyValueStore,
	})
	if err != nil {
		return fmt.Errorf("error creating the key-value store: %s", err)
	}

	return nil
}

func (n *NatsClient) DeleteStream(stream string) error {
	n.logger.Infof("Deleting stream %q", stream)
	err := n.js.DeleteStream(stream)
	return err
}

func (n *NatsClient) DeleteObjectStore(objectStore string) error {
	n.logger.Infof("Deleting object store %q", objectStore)
	err := n.js.DeleteObjectStore(objectStore)
	return err
}

// GetStreamNames returns the list of streams' names.
// The optional param `optFilter` accepts 0 or 1 value.
func (n *NatsClient) GetStreamNames(optFilter ...*regexp.Regexp) ([]string, error) {
	if len(optFilter) > 1 {
		return nil, errors.ErrNoOptFilter
	}

	var regexpFilter *regexp.Regexp
	if len(optFilter) == 1 {
		regexpFilter = optFilter[0]
	}

	streamsChannel := n.js.StreamNames()
	streams := make([]string, 0)

	for streamName := range streamsChannel {
		nameMatchFilter := regexpFilter == nil || regexpFilter.MatchString(streamName)
		if nameMatchFilter {
			streams = append(streams, streamName)
		}
	}

	return streams, nil
}

func (n *NatsClient) getNodesSubjects(nodes entity.NodesStreamConfig) []string {
	subjects := make([]string, 0, len(nodes)*2)

	for _, nodeCfg := range nodes {
		subSubject := nodeCfg.Subject + ".*"
		subjects = append(subjects, nodeCfg.Subject, subSubject)
	}

	return subjects
}
