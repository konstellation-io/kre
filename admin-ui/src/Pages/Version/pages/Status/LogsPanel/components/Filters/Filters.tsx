import GroupSelect, {
  GroupSelectData
} from 'Components/Form/GroupSelect/GroupSelect';
import MultiSelect, {
  MultiSelectOption,
  SelectTheme
} from 'Components/Form/MultiSelect/MultiSelect';

import DateFilter from './components/DatesFilter/DateFilter';
import { GetLogTabs_logTabs_filters } from 'Graphql/client/queries/getLogs.graphql';
import Left from 'Components/Layout/Left/Left';
import LevelIcon from 'Components/LevelIcon/LevelIcon';
import { LogLevel } from 'Graphql/types/globalTypes';
import { NodeSelection } from 'Graphql/client/typeDefs';
import React from 'react';
import Right from 'Components/Layout/Right/Right';
import SearchSelect from 'Components/Form/SearchSelect/SearchSelect';
import cx from 'classnames';
import styles from './Filters.module.scss';
import useWorkflowsAndNodes from 'Hooks/useWorkflowsAndNodes';

// TODO: use GroupSelectData in the filters instead of [string]
function nodesSelectionToDoubleSelector(
  selections: NodeSelection[]
): GroupSelectData {
  const data: { [key: string]: string[] } = {};

  selections.forEach(({ workflowName, nodeNames }) => {
    data[workflowName] = nodeNames;
  });

  return data;
}

function doubleSelectorToNodesSelection(
  selections: GroupSelectData
): NodeSelection[] {
  const data: NodeSelection[] = Object.entries(selections).map(
    ([workflowName, nodeNames]) => ({
      workflowName,
      nodeNames,
      __typename: 'NodeSelection'
    })
  );

  return data;
}

const levelsOrdered = [
  LogLevel.INFO,
  LogLevel.DEBUG,
  LogLevel.WARN,
  LogLevel.ERROR
];

type Props = {
  versionId: string;
  updateFilters: Function;
  filterValues: GetLogTabs_logTabs_filters;
};
function Filters({ updateFilters, filterValues, versionId }: Props) {
  const { workflowsAndNodesNames } = useWorkflowsAndNodes(versionId);

  const levelOptions = levelsOrdered.map(
    (level: LogLevel) =>
      ({
        label: level,
        Icon: <LevelIcon level={level} />
      } as MultiSelectOption<LogLevel>)
  );

  function onNodeSelection(newSelection: GroupSelectData) {
    updateFilters({
      nodes: doubleSelectorToNodesSelection(newSelection)
    });
  }

  function onLevelSelection(newLevels: LogLevel[]) {
    updateFilters({ levels: newLevels });
  }

  function onSearchUpdate(newSearch: string) {
    updateFilters({ search: newSearch });
  }

  return (
    <div className={styles.container}>
      <Left className={cx(styles.filterContainer, styles.leftContainer)}>
        <div className={styles.searchFilter}>
          <SearchSelect
            options={[]}
            onChange={onSearchUpdate}
            value={filterValues.search}
            className={styles.searchForm}
            placeholder="Search"
            hideError
            hideLabel
            showSearchIcon
          />
        </div>
        <div className={styles.selectProcesses}>
          <GroupSelect
            options={workflowsAndNodesNames}
            formSelectedOptions={nodesSelectionToDoubleSelector(
              filterValues.nodes || []
            )}
            onChange={onNodeSelection}
            label=""
            placeholder="Select Processes"
            className={styles.selectProcessForm}
            hideError
            hideSelections
          />
        </div>
      </Left>
      <Right className={cx(styles.filterContainer, styles.rightFilters)}>
        <div className={styles.filterLevel}>
          <MultiSelect<LogLevel>
            onChange={onLevelSelection}
            label=""
            hideError
            placeholder="ALL LEVELS"
            selectionUnit="LEVEL"
            selectAllText="ALL LEVELS"
            options={levelOptions}
            formSelectedOptions={filterValues.levels || []}
            theme={SelectTheme.LIGHT}
          />
        </div>
        <DateFilter
          selectedOption={filterValues.dateOption}
          updateFilters={updateFilters}
          formStartDate={filterValues.startDate}
          formEndDate={filterValues.endDate}
        />
      </Right>
    </div>
  );
}

export default Filters;
