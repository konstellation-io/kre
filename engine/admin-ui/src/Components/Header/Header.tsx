import React, { FunctionComponent, memo, ReactElement } from 'react';

import { GetMe } from 'Graphql/queries/types/GetMe';
import Settings from 'Components/Settings/Settings';
import { get } from 'lodash';
import styles from './Header.module.scss';
import { useQuery } from '@apollo/client';
import ROUTE from 'Constants/routes';
import { Link, useRouteMatch } from 'react-router-dom';
import cx from 'classnames';

import GetMeQuery from 'Graphql/queries/getMe';

type Props = {
  children?: ReactElement | ReactElement[] | null;
  hideSettings?: boolean;
};
const Header: FunctionComponent<Props> = ({
  children,
  hideSettings = false
}) => {
  const { data, loading } = useQuery<GetMe>(GetMeQuery);
  const isAtVersions = useRouteMatch(ROUTE.VERSIONS)?.isExact;

  if (loading)
    return <div className={styles.splash} data-testid="splashscreen" />;

  const username = get(data?.me, 'email', '');

  return (
    <header className={styles.container} data-testid="app-header">
      <Link
        to={ROUTE.HOME}
        className={cx({ [styles.linkDisabled]: isAtVersions })}
      >
        <img
          className={styles.konstellationText}
          src={'/img/brand/konstellation.svg'}
          alt="konstellation"
        />
      </Link>
      <div className={styles.customHeaderElements}>{children}</div>
      {!hideSettings && <Settings label={username} />}
    </header>
  );
};

export default memo(Header);
