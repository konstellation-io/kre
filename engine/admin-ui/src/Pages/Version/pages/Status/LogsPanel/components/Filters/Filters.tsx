import {
  GroupSelect,
  GroupSelectData,
  Left,
  MultiSelect,
  MultiSelectOption,
  MultiSelectTheme,
  Right,
  SearchSelect
} from 'kwc';

import DateFilter from './components/DatesFilter/DateFilter';
import LevelIcon from 'Components/LevelIcon/LevelIcon';
import { LogLevel } from 'Graphql/types/globalTypes';
import { LogPanelFilters, NodeSelection } from 'Graphql/client/typeDefs';
import React from 'react';
import cx from 'classnames';
import styles from './Filters.module.scss';
import useWorkflowsAndNodes from 'Hooks/useWorkflowsAndNodes';

// TODO: use GroupSelectData in the filters instead of [string]
function nodesSelectionToDoubleSelector(
  selections: NodeSelection[]
): GroupSelectData {
  const data: { [key: string]: string[] } = {};

  selections.forEach(({ workflowName, nodeNames }) => {
    data[workflowName] = nodeNames;
  });

  return data;
}

function doubleSelectorToNodesSelection(
  selections: GroupSelectData
): NodeSelection[] {
  const data: NodeSelection[] = Object.entries(selections).map(
    ([workflowName, nodeNames]) => ({
      workflowName,
      nodeNames,
      __typename: 'NodeSelection'
    })
  );

  return data;
}

const levelsOrdered = [
  LogLevel.INFO,
  LogLevel.DEBUG,
  LogLevel.WARN,
  LogLevel.ERROR
];

type Props = {
  versionName: string;
  updateFilters: Function;
  filterValues: LogPanelFilters;
  runtimeId: string;
};
function Filters({ updateFilters, filterValues, versionName, runtimeId }: Props) {
  const { workflowsAndNodesNames } = useWorkflowsAndNodes(versionName, runtimeId);

  const levelOptions = levelsOrdered.map(
    (level: LogLevel) =>
      ({
        label: level,
        Icon: <LevelIcon level={level} />
      } as MultiSelectOption<LogLevel>)
  );

  function onNodeSelection(newSelection: GroupSelectData) {
    updateFilters({
      nodes: doubleSelectorToNodesSelection(newSelection)
    });
  }

  function onLevelSelection(newLevels: LogLevel[]) {
    updateFilters({ levels: newLevels });
  }

  function onSearchUpdate(newSearch: string) {
    updateFilters({ search: newSearch });
  }

  return (
    <div className={styles.container}>
      <Left className={cx(styles.filterContainer, styles.leftContainer)}>
        <div className={styles.searchFilter}>
          <SearchSelect
            options={[]}
            onChange={onSearchUpdate}
            value={filterValues.search}
            className={styles.searchForm}
            placeholder="Search"
            hideError
            hideLabel
            showSearchIcon
          />
        </div>
        <div className={styles.selectProcesses}>
          <GroupSelect
            options={workflowsAndNodesNames}
            formSelectedOptions={nodesSelectionToDoubleSelector(
              filterValues.nodes || []
            )}
            onChange={onNodeSelection}
            label=""
            placeholder="Select Processes"
            className={styles.selectProcessForm}
            hideError
            hideSelections
          />
        </div>
      </Left>
      <Right className={cx(styles.filterContainer, styles.rightFilters)}>
        <div className={styles.filterLevel}>
          <MultiSelect<LogLevel>
            onChange={onLevelSelection}
            label=""
            hideError
            placeholder="ALL LEVELS"
            selectionUnit="LEVEL"
            selectAllText="ALL LEVELS"
            options={levelOptions}
            formSelectedOptions={filterValues.levels || []}
            theme={MultiSelectTheme.LIGHT}
          />
        </div>
        <DateFilter
          selectedOption={filterValues.dateOption}
          updateFilters={updateFilters}
          formStartDate={filterValues.startDate}
          formEndDate={filterValues.endDate}
        />
      </Right>
    </div>
  );
}

export default Filters;
