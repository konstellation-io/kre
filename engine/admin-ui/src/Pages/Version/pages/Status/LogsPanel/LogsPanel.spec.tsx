import Filters from './components/Filters/Filters';
import Header from './components/Header/Header';
import LogsPanel from './LogsPanel';
import React from 'react';
import { shallow } from 'enzyme';
import * as apolloClient from '@apollo/client';

const mocksLogTabs = {
  data: {
    logTabs: [
      {
        runtimeId: 'runtimeIdMock',
        nodeId: 'nodeIdMock',
        nodeName: 'nodeNameMock',
        uniqueId: 'foo'
      }
    ]
  }
};

apolloClient.useQuery = () => mocksLogTabs;
apolloClient.useApolloClient = () => ({ writeData: jest.fn() });
apolloClient.useReactiveVar = () => [];

jest.mock('Graphql/client/cache', () => ({}));

jest.mock('react-router', () => ({
  useLocation: () => ({
    pathname: ''
  }),
}));

describe('Logs', () => {
  let wrapper: any;
  const mockSetSelectedNode = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<LogsPanel />);
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('show right components', () => {
    expect(wrapper.exists('.container')).toBeTruthy();
    expect(wrapper.exists(Header)).toBeTruthy();
  });

  // FIXME: fix it when you can mock or use useEffect with shallow
  it.skip('calls setSelectedNode on shield click', () => {
    expect(wrapper.exists(Filters)).toBeTruthy();

    wrapper.find('.shield').simulate('click');

    expect(mockSetSelectedNode).toHaveBeenCalledTimes(1);
    expect(mockSetSelectedNode).toHaveBeenCalledWith(undefined);
  });
});
