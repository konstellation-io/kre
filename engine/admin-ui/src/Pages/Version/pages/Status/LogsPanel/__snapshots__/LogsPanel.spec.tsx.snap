// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logs matches snapshot 1`] = `
<Fragment>
  <div
    className="container opened hidden"
  >
    <Header
      fullScreen={false}
      opened={Array []}
      toggleFullScreen={[Function]}
      togglePanel={[Function]}
    />
    <TabContainer
      activeTabId={Array []}
      closeTab={[Function]}
      contextMenuActions={
        Array [
          Object {
            "Icon": Object {
              "$$typeof": Symbol(react.memo),
              "compare": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
            },
            "callToAction": [Function],
            "text": "duplicate",
          },
          Object {
            "Icon": Object {
              "$$typeof": Symbol(react.memo),
              "compare": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
            },
            "callToAction": [Function],
            "text": "open in a new tab",
          },
          Object {
            "Icon": Object {
              "$$typeof": Symbol(react.memo),
              "compare": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
            },
            "callToAction": [Function],
            "disabled": true,
            "text": "close others",
          },
          Object {
            "Icon": Object {
              "$$typeof": Symbol(react.memo),
              "compare": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
            },
            "callToAction": [Function],
            "text": "close",
          },
        ]
      }
      onTabClick={[Function]}
      tabs={Array []}
    />
  </div>
  <div
    className="shield"
    onClick={[Function]}
  />
</Fragment>
`;
